File: packages/sequence/c/src/message.c
/*
 * jMax
 * Copyright (C) 1994, 1995, 1998, 1999 by IRCAM-Centre Georges Pompidou, Paris, France.
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * See file LICENSE for further informations on licensing terms.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Based on Max/ISPW by Miller Puckette.
 *
 * Authors: Maurizio De Cecco, Francois Dechelle, Enzo Maggi, Norbert Schnell.
 *
 */
#include "fts.h"
#include "seqsym.h"
#include "sequence.h"
#include "eventtrk.h"

typedef struct _message_
{
  fts_object_t _obj;
  fts_symbol_t s;
  int ac;
  fts_atom_t *at;
  int pos; /* property */
} message_t;

static void
message_set_message(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  message_t *this = (message_t *)o;
  int i;

  if(ac)
    {
      /* implement conventions of conversion from atom list to message */
      if(fts_is_symbol(at))
	{
	  this->s = fts_get_symbol(at);

	  /* skip selector */
	  ac--;
	  at++;
	}
      else if(ac == 1)
	this->s = fts_get_selector(at);
      else
	this->s = fts_s_list;

      if(ac != this->ac)
	{
	  if(this->at)
	    fts_block_free(this->at, this->ac * sizeof(fts_atom_t));
	  
	  this->at = (fts_atom_t *)fts_block_alloc(ac * sizeof(fts_atom_t));
	}
      
      this->ac = ac;
      
      for(i=0; i<ac; i++)
	this->at[i] = at[i];
    }
  else
    {
      this->s = 0;

      if(this->at)
	fts_block_free(this->at, this->ac * sizeof(fts_atom_t));
	  
      this->ac = 0;
    }
}

/**************************************************************
 *
 *  set event methods
 *
 */

static void
message_set(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  message_t *this = (message_t *)o;
  
  this->s = fts_get_symbol(at + 0);
  this->pos = fts_get_int(at + ac - 1);
}

void 
message_get_atoms(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  message_t *this = (message_t *)o;
  int *n = fts_get_ptr(at);
  fts_atom_t *a = fts_get_ptr(at + 1);

  fts_set_symbol(a + 0, this->s);
  fts_set_int(a + 1, this->pos);

  *n = 2;
}

/**************************************************************
 *
 *  class
 *
 */

static void
message_init(fts_object_t *o, int winlet, fts_symbol_t s, int ac, const fts_atom_t *at)
{
  message_t *this = (message_t *)o;
  
  this->s = 0;
  this->ac = 0;
  this->at = 0;

  message_set_message(o, 0, 0, 1, at + 1);

  if(ac > 1)
    this->pos = fts_get_number_int(at + 2);
}

static fts_status_t
message_instantiate(fts_class_t *cl, int ac, const fts_atom_t *at)
{
  fts_class_init(cl, sizeof(message_t), 0, 0, 0); 
  
  fts_method_define_varargs(cl, fts_SystemInlet, fts_s_init, message_init);

  fts_method_define_varargs(cl, fts_SystemInlet, seqsym_set, message_set);
  fts_method_define_varargs(cl, fts_SystemInlet, seqsym_get_atoms, message_get_atoms);

  return fts_Success;
}

void
message_config(void)
{
  fts_class_install(seqsym_message, message_instantiate);
}

File: packages/ispw/java/ircam/jmax/ispw/Message.java
//
// jMax
// Copyright (C) 1994, 1995, 1998, 1999 by IRCAM-Centre Georges Pompidou, Paris, France.
// 
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public License
// as published by the Free Software Foundation; either version 2.1
// of the License, or (at your option) any later version.
// 
// See file COPYING.LIB for further informations on licensing terms.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
// 
// You should have received a copy of the GNU Lesser General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
// 

package ircam.jmax.ispw;

import java.awt.*;
import java.awt.event.*;
import java.util.*;

import ircam.jmax.fts.*;
import ircam.jmax.editors.patcher.*;
import ircam.jmax.editors.patcher.objects.*;
import ircam.jmax.editors.patcher.interactions.*;

//
// The "message box" graphic object.
//

public class Message extends Editable implements FtsMessageListener, FtsIntValueListener
{
  boolean isFlashing = false;

  public Message( FtsGraphicObject theFtsObject)
  {
    super( theFtsObject);
  }

  // ----------------------------------------
  // ``Args'' property
  // ----------------------------------------

  public String getArgs()
  {
    // get the correct String from the object's "Message" property, that may change

    return ((FtsMessageObject)ftsObject).getMessage();
  }


  // ----------------------------------------
  // Text area offset
  // ----------------------------------------
  private static final int TEXT_X_OFFSET = 3;
  private static final int TEXT_Y_OFFSET = 2;

  public int getTextXOffset()
  {
    return TEXT_X_OFFSET;
  }

  public int getTextYOffset()
  {
    return TEXT_Y_OFFSET;
  }

  public int getTextWidthOffset()
  {
    return 8;
  }

  public int getTextHeightOffset()
  {
    return 4;
  }

  public boolean isMultiline()
  {
    return true;
  }

  public void redefine( String text) 
  {
    ((FtsMessageObject)ftsObject).setMessage( text);
  }

  public void redefined()
  {
    messageChanged( ((FtsMessageObject)ftsObject).getMessage());
  }

  // Set the text when FTS change the message content

  public void messageChanged(String message)
  {
    if(isEditing())
      itsSketchPad.abortTextEditing();

    redraw();
    redrawConnections();

    //super.redefine( message);
    computeRenderer();
    renderer.update();

    redraw();
    redrawConnections();
  }

  public void gotSqueack(int squeack, Point mouse, Point oldMouse)
  {          
    // Send a bang message to the system inlet
    if (Squeack.isDown(squeack))
	((FtsMessageObject)ftsObject).sendBang();
  }

  public void valueChanged(int value) 
  {
    if (value != 0)
      isFlashing = true;
    else
      isFlashing = false;

    updateRedraw();
  }

  // ----------------------------------------
  // ``TextBackground'' property
  // ----------------------------------------

  public Color getTextForeground()
  {
      return Color.black;
  }

  public Color getTextBackground()
  {
    if( isFlashing) 
      return Settings.sharedInstance().getUIColor();
    else
      {
	if (isSelected()) 
	  return Settings.sharedInstance().getUIColor().darker();
	else
	  return Color.white;
      }
  }
  
  public void paint(Graphics g) 
  {
    g.setColor(getTextBackground());
    g.fillRect( getX()+1, getY()+1, getWidth()-2, getHeight()-2);
    
    drawContent( g);

    super.paint( g);
  }

  public void updatePaint(Graphics g) 
  {
    g.setColor(getTextBackground());
    g.fillRect( getX()+1, getY()+1, getWidth()-2, getHeight()-2);
    
    drawContent( g);
  }
}







